### PR TITLE
[#19] Improve error handling

### PR DESCRIPTION
**Pull request by daenamkim**
Part of #19 

- [x] Cell side error handling
- [x] Handing UI menu side error handling
- [x] Refactor codes

**Manual Tests**

- Create a new spreadsheet
- Go to `Add-ons > Get add-ons`
- Type `multibaas` in search box
- Install (wait for installing)
- Go to `Add-ons > MultiBaas for Google Sheets > Set Deployment ID` and set it
- Go to `Add-ons > MultiBaas for Google Sheets > Set API Key` and set it
- Check error messages on using UI menu and custom formula in the cell

### CODE REVIEW

File: src/Code.js [hunk 1]
```diff
@@ -151,6 +166,7 @@ function onOpen() {
     .addItem('Delete All Settings', 'deleteAllSettings')
     .addItem('Post to the blockchain', 'postToBlockchain')
     .addItem('Refresh current cell', 'refreshCurrentCell')
+    .addItem('Version', 'showVersion')
```
Comment (shoenseiwaso) [inline src/Code.js:169]: Good idea

**Comment by daenamkim** [comment]
> This all makes sense - will wait to test functionality (and then go through the code again), pending your thoughts on the best way to test it out.

Here are test cases to see error messages

Deployment ID and API Key: https://docs.google.com/spreadsheets/d/1g7lodtQvJLeweF8LsHE9RI31cmOOhxTiOBNEWUypRfE/edit#gid=2104841724

**Without Credentials**

- Put formula in the cell
`=MBTX("0xfe9e4b800d14c36f2e8c26ab44ffcfcbf55ac71d6f0d5f2ac95b3d63c7f71569")`

Screen Shot 2020-08-06 at 12 19 58

- `Add-on > MultiBaas for Google Sheets > Set deployment ID`
- Delete the formula and put `=MBTX("0xfe9e4b800d14c36f2e8c26ab44ffcfcbf55ac71d6f0d5f2ac95b3d63c7f71569")` again

Screen Shot 2020-08-06 at 12 20 45

- `Add-on > MultiBaas for Google Sheets > Delete all settings`
- `Add-on > MultiBaas for Google Sheets > Refresh current cell` and `Post to the blockchain` with any cell

Screen Shot 2020-08-06 at 12 08 13

**With Credentials**

You need to set Deployment ID and API key

- `Add-on > MultiBaas for Google Sheets > Refresh current cell`

Screen Shot 2020-08-06 at 11 38 10

- `Add-on > MultiBaas for Google Sheets > Post to the blockchain` with any cell

Screen Shot 2020-08-06 at 11 37 38

URL in the dialog could be replaced depending on MultiBaas for Google Sheets page.

File: src/Code.js [hunk 2]
```diff
@@ -2,32 +2,7 @@
...
-  } else {
-    testProperties[key] = value;
-  }
-}
-
-function getProperty(key) {
-  const properties = PropertiesService.getDocumentProperties();
-  return properties ? properties.getProperty(key) : testProperties[key];
-}
+const VERSION = '1.0.0';
```
Comment (shoenseiwaso) [inline src/Code.js:5]: Can we use dynamic logic to either be the `git` tag, or the `git` tag & the latest commit? Something like:
```
$(git describe --tags --match 'v*[0-9].*[0-9].*[0-9]')
```

We can populate at build time.

File: src/Code.js [hunk 3]
```diff
@@ -58,27 +38,47 @@ function setApiKey() {
...
 function deleteAllSettings() {
-  const properties = PropertiesService.getDocumentProperties();
-  if (properties) {
-    properties.deleteAllProperties();
-  } else {
-    testProperties = {};
+  const ui = SpreadsheetApp.getUi();
+  const result = ui.alert(
+    'Are you sure to delete all settings?',
+    ui.ButtonSet.YES_NO,
```
Comment (shoenseiwaso) [inline src/Code.js:54]: I know it's already `deleteAllSettings()` from a previous PR, but let's change this to be `resetCredentials()` and `"Are you sure you want to reset the Deployment ID and API key?"`, since that's the only/primary use for this at the moment.

File: src/Code.js [hunk 4]
```diff
@@ -58,27 +38,47 @@ function setApiKey() {
...
+  );
+
+  if (result === ui.Button.YES) {
+    deleteAllProperties();
+    ui.alert('Deployment ID and API key have been deleted');
```
Comment (shoenseiwaso) [inline src/Code.js:59]: Is there a reason to have another dialogue popup for this (and other) menu items? Perhaps a popup only in the case of failure?

File: src/Code.js [hunk 5]
```diff
@@ -58,27 +38,47 @@ function setApiKey() {
...
   }
-  SpreadsheetApp.getUi()
-    .alert('Deployment ID and API key have been removed.');
 }
 
 function refreshCurrentCell() {
+  const ui = SpreadsheetApp.getUi();
+
+  if (!credentialsExist()) {
+    ui.alert('Credentials are invalid. Please see https://www.curvegrid.com/multibaas-for-google-sheets/credentials');
```
Comment (shoenseiwaso) [inline src/Code.js:67]: These URLs don't exist yet - let's chat about how to handle them/where to put them (docs vs. landing page, etc.).

File: src/Code.js [hunk 6]
```diff
@@ -58,27 +38,47 @@ function setApiKey() {
...
   const text = result.getResponseText();
   if (button === ui.Button.OK) {
-    setProperty(PROP_MB_API_KEY, text);
-    ui.alert(`API key is ${text}.`);
+    try {
+      setProperty(PROP_MB_API_KEY, text);
+      ui.alert(`API key is ${text}`);
+    } catch (e) {
+      ui.alert(e.message);
+    }
```
Comment (shoenseiwaso) [inline src/Code.js:46]: I know this was also done in a separate PR, but is there a way to set the Deployment ID and API in a single popup, as a convenience? Or not available in Google Sheets?

File: src/Code.js [hunk 7]
```diff
@@ -140,6 +150,11 @@ function postToBlockchain() {
   }
 }
 
+function showVersion() {
+  SpreadsheetApp.getUi()
+    .alert(`MultiBaas for Google Sheets Version ${VERSION}`);
+}
+
```
Comment (shoenseiwaso) [inline src/Code.js:157]: On further thought, let's show the deployment ID here as well. It could be very useful to help keep track of which sheet is using which deployment.

File: src/library/Validate.js [hunk 8]
```diff
@@ -36,22 +57,16 @@ function validateBlockNumOrHash(numOrHash) {
...
-  }
-
-  if (hash.substring(0, 2).toLowerCase() !== '0x') {
-    throw new Error("hash must start with '0x'");
-  }
-
-  if (hash.length !== 66) {
-    throw new Error(`invalid hash length of ${hash.length}, should be 64 hex characters long excluding the '0x' prefix`);
+function validateDeploymentId(deploymentId) {
+  if (!deploymentId || !RegExp('^[a-z0-9]+$', 'i').test(deploymentId)) {
```
Comment (shoenseiwaso) [inline src/library/Validate.js:61]: We have deployment IDs that include other characters (dashes, etc.), and I think we should also allow full URLs as well (by default, `multibaas.com` is appended).

File: src/Code.js (hunk 4 above)
Comment (daenamkim) [reply src/Code.js:59]: It is to let users know what they have done actually but it is ok to remove it when you feel not so useful :)

File: src/Code.js (hunk 5 above)
Comment (daenamkim) [reply src/Code.js:67]: Yeah, I am thinking of using https://github.com/curvegrid/v2.curvegrid.com/pull/198 after merged but we can chat later. :)

File: src/Code.js (hunk 6 above)
Comment (daenamkim) [reply src/Code.js:46]: To be honest, it was my lazy I will check this out like a form  🔍

Comment (daenamkim) [reply src/Code.js:46]: Can I create a follow-up ticket for this? because I am testing custom dialog with html but:

- [Need to add another permission scope](https://developers.google.com/apps-script/reference/base/ui#showModalDialog(Object,String)): `https://www.googleapis.com/auth/script.container.ui` it seems not related to our add-on features.
image 
- Create custom CSS to keep style consistency or make all dialogs with custom dialogs >.<;;

What do you think?

Comment (shoenseiwaso) [reply src/Code.js:46]: Yes, let's do as a follow-on!

File: src/Code.js (hunk 4 above)
Comment (shoenseiwaso) [reply src/Code.js:59]: Ok got it. Let's remove it across the add-on.

File: src/Code.js (hunk 6 above)
Comment (daenamkim) [reply src/Code.js]: Follow-up ticket: https://github.com/curvegrid/multibaas-for-google-sheets/issues/27

File: src/library/Validate.js (hunk 8 above)
Comment (daenamkim) [reply src/library/Validate.js]: I just decided to use deployment's host (it is ok to copy full URL path) only for ID because to:

- avoid incorrect copy
- make it simple

**Review by shoenseiwaso — APPROVED** [review]
LGTM!
